src/Common/Auth/AuthUtils.php

```diff
@@ -967,4 +961,55 @@ private function clearFromMemory(&$password)
             $password = '';
         }
     }
+
+    /**
+     * Validates a google ID token and returns a user on success. If validation
+     * fails, return false.
+     *
+     * @param $token
+     * @return array|bool
+     */
+    static public function verifyGoogleSignIn($token)
+    {
+        $return = false;
+        if (!empty($token)) {
+            $client = new Google_Client([
+                'client_id' => $GLOBALS['google_signin_client_id']]);  // Specify the CLIENT_ID of the app that accesses the backend
+            $payload = $client->verifyIdToken($token);
+            if ($payload) {
+                //See if email address exists in user table
+                $user = UserService::getUserByGoogleSigninEmail($payload['email']);
+                if ($user !== false) {
+                    $_SESSION['GoogleSignInPayload'] = $payload;
+                    $_SESSION['user_logged_in_with_google'] = true;
+                    $_SESSION['google_signin_token'] = $token;
+                    $return = $user;
+                }
+            }
+        }
+
+        return $return;
+    }
+
+    /**
+     * Given an associative array representing the user, set the session variables
+     * @param $username
+     * @param $hash
+     * @param array $userInfo
+     * @param $authGroup
+     */
+    public static function setSessionVariables($username, $hash, $userInfo, $authGroup)
```

Comment:
would call this `setUserSessionVariables` to avoid confusion with the patient auth mechanism

Reply:
Rebased it